init.sls

```diff
@@ -72,20 +72,31 @@ chown_pgdata:
 
 # Deploy users. Sort them, so order is not changing between salt runs
 {% for config in pillar['postgresql']['users']|default({}) %}
-createuser-{{ loop.index }}:
+createuser-{{ config['username'] }}:
   cmd.run:
     - name: psql -t -c "CREATE ROLE {{ config['username'] }} {{ config['options']|default('NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN') }};"
     - unless: psql -t -c "SELECT 1 FROM pg_roles WHERE rolname='{{ config['username'] }}'" |grep -q 1
     - runas: postgres
+{% endfor %}
 
-# The "replication" keyword is not a real database, but just used for permissions in pg_hba.conf
+{% for config in pillar['postgresql']['databases']|default({}) %}
 {% if config['database'] != "replication" %}
-createdb-{{ loop.index }}:
+createdb-{{ config['database'] }}:
   cmd.run:
     - name: psql -t -c "CREATE DATABASE {{ config['database'] }} OWNER {{ config['username'] }};"
     - unless: psql -t -c "SELECT 1 FROM pg_database WHERE datname='{{ config['database'] }}'" |grep -q 1
     - runas: postgres
     - require:
-      - cmd: createuser-{{ loop.index }}
+      - cmd: createuser-{{ config['username'] }}
 {% endif %}
 {% endfor %}
+
+{% for config in pillar['postgresql']['schemas']|default({}) %}
+createschema-{{ config['schemaname'] }}:
```

Comment:
If I understand correctly, a schema always depends on a database. Therefore, a schema could be sub-section of the database configuration?

Something like this:

```yaml
databases:
  - database: mydb
    owner: myuser
    schema: myschema
```

You are also hard-depending on a username when creating a schema, so it might even be more consise to migrate both to the user configuration section. Maybe the user section then might be a little cluttered - we should evaluate both suggestions.

Reply:
Yes, definitely. In addition to your suggestion, a schema owner:
`
databases:
      - database: mzdb
         owner: myowner
         schemas:
          - schemaname: myschema
             owner: myuser
         extensions: [foo, bar]
`